### DISCUSSION TITLE
[PHP/Laravel] General Error: 2014 Cannot execute queries while other unbuffered queries are active

### DISCUSSION THREAD

**Discussion by ranbole** [planetscale/discussion#546] [Q&A] [answered]
Hi There,

I am in the process of switching my DB from Google's Cloud SQL to Planetscale. 

And I've encounted the following error during testing:

`SQLSTATE[HY000]: General error: 2014 Cannot execute queries while other unbuffered queries are active.  Consider using PDOStatement::fetchAll().  Alternatively, if your code is only ever going to run against mysql, you may enable query buffering by setting the PDO::MYSQL_ATTR_USE _BUFFERED_QUERY attribute.`

This error tends to happen when I am looping through a dataset in my app and continuisly doing reads and writes to the DB through Laravel's ORM. 

I've searched online and have taken the following actions:

- Added  `PDO::MYSQL_ATTR_USE_BUFFERED_QUERY => true` in Laravel DB config:
- Updated my collation in Laravel DB config to `utf8mb4_0900_ai_ci` to match that is set on Planetscale's DB, and updated all my tables to the same collation

However I am still getting this General Error 2014,

I've attached my Laravel database config below.

Note that I have been running my app on Google's Cloud SQL for a few years and this issue has never happened. So the app code should be fine here.

Is this a known issue with Planescale/Vitess accessed through Laravel's ORM or PDO? 

If not anyone got any ideas how to fix?

Thanks 

```
'mysql' => [
          'driver' => 'mysql',
          'host' => env('DB_HOST', '127.0.0.1'),
          'port' => env('DB_PORT', '3306'),
          'database' => env('DB_DATABASE', 'forge'),
          'username' => env('DB_USERNAME', 'forge'),
          'password' => env('DB_PASSWORD', ''),
          'unix_socket' => env('DB_SOCKET', ''),
          'charset' => 'utf8mb4',
          'collation' => 'utf8mb4_0900_ai_ci',
          'prefix' => '',
          'prefix_indexes' => true,
          'strict' => true,
          'engine' => null,
          'options' => extension_loaded('pdo_mysql') ? array_filter([
              PDO::MYSQL_ATTR_SSL_CA => env('MYSQL_ATTR_SSL_CA'),
              PDO::ATTR_EMULATE_PREPARES => true,
              PDO::MYSQL_ATTR_LOCAL_INFILE => true,
              PDO::MYSQL_ATTR_USE_BUFFERED_QUERY => true
          ]) : [],
      ],
```

**Comment by orware** [comment]
Hello @ranbole,

Could you please send in a support request over to https://support.planetscale.com/ so we can look into this together more in depth?

There have been several other users that have reported encountering the same issue over the last year, with the `collation` adjustment you mentioned being helpful in a few of those cases.

Most of the other occurrences were sporadic as well and it's possible there may be an underlying error / query issue that is also contributing here as well so it'll be useful to be able to look up the details for your database on our end after receiving your ticket and provide additional suggestions for you.

It shouldn't be related to the error as far as I know, but since I noticed the `PDO::MYSQL_ATTR_LOCAL_INFILE` option was enabled in your configuration example above I just wanted to mention that it likely won't have an effect in the Vitess + MySQL environment for your database as `LOAD LOCAL INFILE` isn't something that we currently allow/support [at the Vitess level](https://github.com/vitessio/vitess/issues/2976).

**Reply by orware** [reply]
Hello @ranbole,

I don't believe I've seen a ticket arrive yet about your issue, but another tip that has recently come up for this topic that may also be helpful for you is that the error can occur in relation to attempting to write a large amount of data to a `TEXT` column that is greater than its max size.

Since you mentioned the following:

> I am looping through a dataset in my app and continuously doing reads and writes to the DB through Laravel's ORM.

I would explore whether any of the writes that are occurring include any columns that may be defined as `TEXT` ones with a large amount of text since this would be a good indicator that would explain what is causing the problem. Increasing the column size to `MEDIUMTEXT` for the problematic column(s) would probably be the most immediate solution in that situation and then continue monitoring afterward.

**Comment by ranbole** [comment]
@orware Thanks for reply! I will go on to submit a request right now.

To your point about `TEXT`, yes the table I'm operating on does have a few `TEXT` cols, but I am certain none of the data I've tried to insert to the table has a text that exceeds the limit as they work fine on a Cloud SQL DB with the same table structure. Most of the time the error happens when trying to execute a `SELECT` statement through Laravel ORM, and `SELECT` statement is the most prominet statement type in the process.

**Reply by orware** [reply]
@ranbole,

Thank you for filing that ticket! I'll be continuing the discussion over there and if we come up with a solution together I'll post back here with that for others to be able to see.

**Comment by orware** [comment, marked as answer]
For others coming here with a similar issue, the end result that was found is that this issue relates to the following PHP bug report:
https://bugs.php.net/bug.php?id=78797

In short, for lengthier queries, responses from PlanetScale databases can exceed PHP's default 4096 byte length for the `mysqlnd.net_cmd_buffer_size` value.

This isn't really a problem in a regular MySQL environment as the message returned to PHP when there is a query error tends to be fairly short, for example:
 
```
Data too long for column '<EXAMPLE_COLUMN>' at row 1
```

But in our Vitess + MySQL environment, a lengthier error message is included, which can then result in that 4096 bytes being exceeded if it is a longer query. 

For example:

```
Data too long for column '<EXAMPLE_COLUMN>' at row 1 (errno 1406) (sqlstate 22001) (CallerID: <USERNAME>): Sql: "update <TABLE_NAME> set <EXAMPLE_COLUMN> = :v1, <EXAMPLE_COLUMN_2> = :v2,<EXAMPLE_COLUMN_3> = :v3 where id = :v4 ...
 ```

In addition to receiving the unbuffered query message, user's may also encounter this one as well from PHP:
 
```
PDOStatement::execute(): Error reading result set's header
```

Which unfortunately masks the true error message that they should be seeing.
 
In order to remedy this, adding a call to increase the value would be needed and the following `16384` size should be good enough for most purposes, but the size of the value itself will be dependent on what your largest query might be that could run into an error:

```
ini_set('mysqlnd.net_cmd_buffer_size', 16384);
```

In order to fix this from a Laravel perspective, following the middleware tip and then adding to Laravel's Kernel file as mentioned here seems like it might be a good approach to utilize:
https://laracasts.com/discuss/channels/laravel/over-written-phpini-in-laravel
 
But essentially anywhere that's a key location within Laravel so you know that the setting is pretty much always set should work well, or even adjusting the setting within the `php.ini` file directly can be done too.
 
I feel like this explanation does a much better job overall to demonstrate why this issue has been fairly specific to PHP so far, as other MySQL clients may not impose a similar restriction on that response buffer.
 
In the future, [adjustments within Vitess](https://github.com/vitessio/vitess/pull/12899) may change things so that not as much data gets returned in those error packets as they do today, which would be another way of avoiding the negative impact for PHP users, but hopefully the above tip will prove useful for any other users experiencing a similar issue.

**Reply by ranbole** [reply]
Thank you for working with me to resolve the issue @orware ! :+1:

**Comment by jamesmills** [comment]
I'm also have issues with this. I've tried the suggested solution but I'm still seeing the error.

```
SQLSTATE[HY000]: General error: 2014 Cannot execute queries while other unbuffered queries are active.  Consider using PDOStatement::fetchAll().  Alternatively, if your code is only ever going to run against mysql, you may enable query buffering by setting the PDO::MYSQL_ATTR_USE_BUFFERED_QUERY attribute.
```

and

```
PDOStatement::execute(): Error reading result set's header
```

I'm using Laravel on Laravel Vapor with Planetscale Scaler Pro.

I'm not doing anything fancy. This seems to be an issue with an UPDATE command that's running on a queued job after a few other queries have run. 

Any ideas?

**Reply by samlambert** [reply]
Have you tried pointing it at a local MySQL and see if it does the same thing? I think this might be a client issue.

**Reply by jamesmills** [reply]
When running this locally I'm using a local MySQL database and I'm not having this issue. The issue in production seems to be connected with the fact I'm running some tasks on queued jobs which obviously can all operate close to "at the same time". This seems to be the issue with...

> The error you're encountering, SQLSTATE[HY000]: General error: 2014 Cannot execute queries while other unbuffered queries are active, typically occurs when you're using PDO (PHP Data Objects) to interact with a MySQL database and you attempt to execute a new query before the previous unbuffered query has been fully processed. This can happen if you haven't fetched all rows from the previous query before starting a new one.

This is all a little over my head.

**Reply by orware** [reply]
@jamesmills,

It looks like your initial issue was solved by implementing the solution above which helped to reveal the true error message you were receiving, which we received your separate ticket linking to (https://twitter.com/mattrobenolt/status/1759515282694197368) earlier this morning. I'll be continuing the discussion there on that topic for you shortly.

**Reply by jamesmills** [reply]
Hi @orware, thanks for getting back to me here, on Twitter and following up on the support ticket.

For those finding this thread with a similar issue around the "Cannot execute queries while other unbuffered queries are active." and "Error reading result set's header" I thought I'd update with a little more info about the fix.

I'm using Laravel Vapor along with Planetscale. I was unable to fix this using the ini_set middleware solution. However, I did [find a section in the Vapor Docs](https://docs.vapor.build/projects/environments.html) which suggested this:

> To customize core php.ini directives on our native runtimes, create a php.ini file in the php/conf.d directory of your project with the desired directives.

So I created a new file in `{project_root}/php/conf.d/php.ini` and added `mysqlnd.net_cmd_buffer_size = 16384`.

This fixed the issues I was having with the unbuffered stuff which ultimately helped get to the real issue I was having. 

Hope this helps someone.

Thanks to the Planetscale team for helping me through this.
